pkg-manager/resolve-dependencies/src/resolveDependencyTree.ts

```diff
@@ -234,6 +238,28 @@ export async function resolveDependencyTree<T> (
   const { pkgAddressesByImporters, time } = await resolveRootDependencies(ctx, resolveArgs)
   const directDepsByImporterId = zipObj(importers.map(({ id }) => id), pkgAddressesByImporters)
 
+  let newDefaultCatalogs: Record<string, ResolvedCatalogEntry> | undefined
+  for (const directDependencies of pkgAddressesByImporters) {
+    for (const directDep of directDependencies as PkgAddress[]) {
+      const { alias, normalizedBareSpecifier, version, saveCatalog } = directDep
+      const existingCatalog = opts.catalogs?.default?.[alias]
+      if (existingCatalog != null) {
+        if (existingCatalog !== normalizedBareSpecifier) {
+          globalWarn(
+            `Skip adding ${alias} to the default catalog because it already exists as ${existingCatalog}, please use \`pnpm update\` to update the catalogs.`
+          )
+        }
+      } else if (saveCatalog && normalizedBareSpecifier != null && version != null) {
+        newDefaultCatalogs ??= {}
+        newDefaultCatalogs[alias] = {
+          specifier: normalizedBareSpecifier,
+          version,
```

Comment:
I am not sure this will always work correctly. What if the dependency is installed from git? Or some other non-registry source?

Reply:
I have a test specifically for this case: https://github.com/pnpm/pnpm/blob/c57779d3bbc4f186d03cde59242fee8e83845fb6/pkg-manager/plugin-commands-installation/test/saveCatalog.ts#L85

I am not sure if pnpm can install a git dependency that lacks a version though.

Reply:
That's not how I would expect it to work. If you check in `importers` version will be `version: https://codeload.github.com/kevva/is-positive/tar.gz/97edff6`.

But this is how it already works. I don't know why. I'd expect them to match.

Reply:
> `https://codeload.github.com/kevva/is-positive/tar.gz/97edff6`

What field store this value? `pkgId`? Is the a function for the lockfile to determine which `version` to store?

Reply:
That value is actually only available at a later stage, after peer dependencies resolution. It is pkgId but without the alias. Let's keep it as it is for now.

Reply:
> Let's keep it as it is for now.

It still produce incorrect lockfile though. Do you mean you are going to handle it?

Reply:
No, it seems to work the same as it works on the main branch now.

Reply:
The main branch now saves `version` as actual version?

Reply:
yes

Reply:
The resolved `version` won't need to be saved to the `pnpm-workspace.yaml`. I think `newDefaultCatalogs` can be the `Catalogs` type from `@pnpm/catalog.types` if we generalize `newDefaultCatalogs` to `newCatalogs` as separately suggested above.

Reply:
> The resolved `version` won't need to be saved to the `pnpm-workspace.yaml`.

We do need to save to `newLockfile` whose `catalogs` contains `version`.